File: android/app/src/main/kotlin/games/yandex/wrap/config/AppConfig.kt
package games.yandex.wrap.config

import java.net.URI

data class HttpDefaults(
    val userAgent: String,
    val acceptLanguage: String = "en-US,en;q=0.9",
)

data class YandexEndpoints(
    val platform: String = "android_other",
    val clientWidth: Int = 412,
    val clientHeight: Int = 915,
) {
    fun origin(): URI = URI("https://$YANDEX_HOST")
    fun gamesHome(): URI = URI("https://$YANDEX_HOST/games/")
    fun passportOrigin(): URI = URI("https://passport.$YANDEX_HOST")
    fun authUrl(): URI = URI("${passportOrigin()}/auth?retpath=${gamesHome().toString().urlEncoded()}")
    fun gameUrl(appId: Long): URI = URI("https://$YANDEX_HOST/games/app/$appId")
    fun feedApi(): URI = URI("https://$YANDEX_HOST/games/api/catalogue/v2/feed/")
    fun searchApi(): URI = URI("https://$YANDEX_HOST/games/api/catalogue/v2/search/")
    fun similarApi(): URI = URI("https://$YANDEX_HOST/games/api/catalogue/v2/similar_games/")
    fun gameDetailApi(): URI = URI("https://$YANDEX_HOST/games/api/catalogue/v2/get_game")
    fun tagsApi(): URI = URI("https://$YANDEX_HOST/games/api/catalogue/v2/tags/")
    fun userInfoApi(): URI = URI("https://$YANDEX_HOST/games/api/catalogue/v2/user_info")

    fun isGamesUrl(url: String): Boolean =
        url.startsWith(gamesHome().toString())

    fun documentStartOrigins(): Set<String> = setOf(
        "https://$YANDEX_HOST",
    )

    fun logBridgeOrigins(): Set<String> = documentStartOrigins() + setOf(
        passportOrigin().toString(),
        "https://*.games.s3.yandex.net",
        "https://*.cdn.games.yandex.net",
        "https://*.gamecdn.yandex.net",
        "https://*.game-static.ru",
        "https://game-static.ru",
    )

    fun cookieOrigins(): List<String> = listOf(origin().toString(), passportOrigin().toString())

    private companion object {
        const val YANDEX_HOST = "yandex.ru"
    }
}

data class AppConfig(
    val yandex: YandexEndpoints,
    val http: HttpDefaults,
) {
    companion object {
        fun defaultForLocale(language: String = ""): AppConfig {
            return AppConfig(
                yandex = YandexEndpoints(),
                http = HttpDefaults(
                    userAgent = "Mozilla/5.0 (Linux; Android 13; Pixel 7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/147.0.0.0 Mobile Safari/537.36",
                ),
            )
        }
    }
}

private fun String.urlEncoded(): String =
    java.net.URLEncoder.encode(this, Charsets.UTF_8.name())

File: ios/UGames/Catalog/AppConfig.swift
import Foundation

struct HTTPDefaults {
    let userAgent: String
    let acceptLanguage: String

    static let ios = HTTPDefaults(
        userAgent: "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1",
        acceptLanguage: "en-US,en;q=0.9"
    )
}

struct YandexEndpoints {
    let platform: String
    let clientWidth: Int
    let clientHeight: Int

    init(
        platform: String = "ios",
        clientWidth: Int = 390,
        clientHeight: Int = 844
    ) {
        self.platform = platform
        self.clientWidth = clientWidth
        self.clientHeight = clientHeight
    }

    func origin() -> URL {
        URL(string: "https://\(Self.yandexHost)")!
    }

    func gamesHome() -> URL {
        URL(string: "https://\(Self.yandexHost)/games/")!
    }

    func passportOrigin() -> URL {
        URL(string: "https://passport.\(Self.yandexHost)")!
    }

    func passportAuthURL() -> URL {
        var components = URLComponents(url: passportOrigin().appendingPathComponent("auth"), resolvingAgainstBaseURL: false)!
        components.queryItems = [
            URLQueryItem(name: "retpath", value: gamesHome().absoluteString),
        ]
        return components.url!
    }

    func feedApi() -> URL {
        URL(string: "https://\(Self.yandexHost)/games/api/catalogue/v2/feed/")!
    }

    func searchApi() -> URL {
        URL(string: "https://\(Self.yandexHost)/games/api/catalogue/v2/search/")!
    }

    func similarApi() -> URL {
        URL(string: "https://\(Self.yandexHost)/games/api/catalogue/v2/similar_games/")!
    }

    func gameDetailApi() -> URL {
        URL(string: "https://\(Self.yandexHost)/games/api/catalogue/v2/get_game")!
    }

    func tagsApi() -> URL {
        URL(string: "https://\(Self.yandexHost)/games/api/catalogue/v2/tags/")!
    }

    func userInfoApi() -> URL {
        URL(string: "https://\(Self.yandexHost)/games/api/catalogue/v2/user_info")!
    }

    func gameUrl(_ appId: Int64) -> URL {
        URL(string: "https://\(Self.yandexHost)/games/app/\(appId)")!
    }

    func isGamesUrl(_ url: String) -> Bool {
        url.hasPrefix(gamesHome().absoluteString)
    }

    func cookieOrigins() -> [String] {
        [origin().absoluteString, passportOrigin().absoluteString]
    }

    private static let yandexHost = "yandex.ru"
}

extension YandexEndpoints {
    static let live = YandexEndpoints()
}

struct AppConfig {
    let yandex: YandexEndpoints
    let http: HTTPDefaults
}

extension AppConfig {
    static func live(language: String = Locale.preferredLanguages.first ?? Locale.current.identifier) -> AppConfig {
        AppConfig(yandex: .live, http: .ios)
    }
}
